lib/AST/ArchetypeBuilder.cpp

```diff
@@ -1013,52 +1076,66 @@ bool ArchetypeBuilder::addConformanceRequirement(PotentialArchetype *PAT,
+  // requirement signature, but inferred from this conformance otherwise.
+  auto Kind =
+      Source.getKind() == RequirementSource::ProtocolRequirementSignatureSelf
+          ? RequirementSource::Explicit
+          : RequirementSource::Protocol;
 
-  // Add requirements for each of the associated types.
-  // FIXME: This should use the generic signature, not walk the members.
-  for (auto Member : Proto->getMembers()) {
-    if (auto AssocType = dyn_cast<AssociatedTypeDecl>(Member)) {
-      // Add requirements placed directly on this associated type.
-      auto AssocPA = T->getNestedType(AssocType, *this);
-      // Requirements introduced by the main protocol are explicit in a protocol
-      // requirement signature, but inferred from this conformance otherwise.
-      auto Kind = Source.getKind() ==
-                          RequirementSource::ProtocolRequirementSignatureSelf
-                      ? RequirementSource::Explicit
-                      : RequirementSource::Protocol;
-
-      if (AssocPA != T) {
-        if (addAbstractTypeParamRequirements(AssocType, AssocPA, Kind, Visited))
-          return true;
-      }
+  // Use the requirement signature to avoid rewalking the entire protocol.  This
+  // cannot compute the requirement signature directly, because that may be
+  // infinitely recursive: this code is also used to construct it.
+  if (Proto->isRequirementSignatureComputed()) {
+    auto reqSig = Proto->getRequirementSignature();
 
-      continue;
+    for (auto req : reqSig->getRequirements()) {
+      RequirementSource InnerSource(Kind, Source.getLoc());
+      addRequirement(req, InnerSource, T, Visited);
     }
+  } else {
+    // Conformances to inherit protocols are explicit in a protocol requirement
+    // signature, but inferred from this conformance otherwise.
+    auto InnerKind =
+        Source.getKind() == RequirementSource::ProtocolRequirementSignatureSelf
+            ? RequirementSource::Explicit
+            : RequirementSource::Redundant;
+    RequirementSource InnerSource(InnerKind, Source.getLoc());
+    // Add all of the inherited protocol requirements, recursively.
+    if (auto resolver = getLazyResolver())
+      resolver->resolveInheritedProtocols(Proto);
+
+    for (auto InheritedProto :
+         Proto->getInheritedProtocols(getLazyResolver())) {
+      if (Visited.count(InheritedProto))
+        continue;
+      if (addConformanceRequirement(T, InheritedProto, InnerSource, Visited))
+        return true;
+    }
+
+    // Add requirements for each of the associated types.
+    for (auto Member : Proto->getMembers()) {
+      if (auto AssocType = dyn_cast<AssociatedTypeDecl>(Member)) {
+        // Add requirements placed directly on this associated type.
+        auto AssocPA = T->getNestedType(AssocType, *this);
+
+        if (AssocPA != T) {
+          if (addAbstractTypeParamRequirements(AssocType, AssocPA, Kind,
```

Comment:
`addAbstractTypeParamRequirements` has a hack that could probably go away now:

  // If this is an associated type that already has an archetype assigned,
  // use that information.
  auto *genericEnv = decl->getDeclContext()
      ->getGenericEnvironmentOfContext();
  if (isa<AssociatedTypeDecl>(decl) && genericEnv != nullptr) {

because the requirement signature should already handle all of the cases where this would kick in.